### PR TITLE
fix: remove deletion approved annotation when sync is completed

### LINKED ISSUE #27570: argocd.argoproj.io/deletion-approved is not removed from application after sync is done
Checklist:

- [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
- [x] I've included steps to reproduce the bug.
- [x] I've pasted the output of `argocd version`.

**Describe the bug**

When application has child resources with Delete=confirm sync option set, and resources got deleted, annotation stays on the parent application and it seems it never gets removed. This causes next syncs needing confirmation auto-approved.

**To Reproduce**

1. Create application with autosync & autoprune enabled
2. One of child resources should have Delete=confirm sync option set through annotation
3. Remove this resource from source of truth. Argo CD blocks sync and waits for confirmation - this is ok
4. Click "Confirm Prunning" button -> annotation `argocd.argoproj.io/deletion-approved` is added to the app 
5. child resource is deleted, but annotation stays
6. add add back deleted resource to source of truth
7. remove this resource from git/source of truth again
8. it gets autoconfirmed, because there is already old annotation

**Expected behavior**

IMHO `argocd.argoproj.io/deletion-approved` annotation should be added to the child resources waiting for deleteion and not on the application level

**Screenshots**

N/A

**Version**

```shell
3.3.8
```

**Logs**

N/A

### PR DESCRIPTION
**Pull request by ppapapetrou76**
Closes #27570 

1. Implemented the controller fix to clear stale argocd.argoproj.io/deletion-approved after confirmed delete/prune flows so old approvals are not reused.
2. Added unit-test coverage for the newly introduced deletion-approval annotation cleanup logic in the controller.
3.  Added e2e assertions to verify annotation cleanup and that a new confirm-required sync blocks again until re-approved.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: test/e2e/app_management_test.go [hunk 1]
```diff
@@ -3155,6 +3155,33 @@ func TestDeletionConfirmation(t *testing.T) {
 		ExpectConsistently(OperationPhaseIs(OperationRunning), time.Second, 5*time.Second).
 		When().ConfirmDeletion().
 		Then().Expect(OperationPhaseIs(OperationSucceeded)).
+		And(func(app *Application) {
+			assert.Empty(t, app.Annotations[AnnotationDeletionApproved])
```
Comment (agaudreault) [inline test/e2e/app_management_test.go:3159]: I don't understand the bug. What is the problem with the annotation being left there? 

The annotation contains a timestamp. The timestamp must be greater then the sync start time or the application deletion time (in case we are confirming that an app is deleted).

In both of these case, an old annotation will have no effect.

**Review by agaudreault — CHANGES_REQUESTED** [review]

File: test/e2e/app_management_test.go (hunk 1 above)
Comment (ppapapetrou76) [reply test/e2e/app_management_test.go:3159]: @agaudreault, the problem is that if the annotation is left in the application (after the sync is completed), then during the next sync, resources with Delete=confirm will not prompt for a prune confirmation, but they will be deleted because of the previously left annotation. 

Does it make sense?

Comment (agaudreault) [reply test/e2e/app_management_test.go:3159]: > resources with Delete=confirm will not prompt for a prune confirmation, but they will be deleted because of the previously left annotation.

They should **not** since the timestamp of the last approved will be before the sync startedAt timestamp. Afaik, there are e2e covering this exact usecase.

Comment (ppapapetrou76) [reply test/e2e/app_management_test.go:3159]: @agaudreault I did a deep dive into the code, and your comment is largely correct, but there's a real subtle edge case that the user might have fallen into

the annotation is set with `metav1.Now().Format(time.RFC3339)` which has a precision to the second. 
So if a confirmation happens at the end of the second, and the next auto-sync triggers within the same second then the approval is missing. 

This is indeed rare and I believe that the e2e [test](https://github.com/argoproj/argo-cd/blob/f40bc68440a57610488ae7a3bf0b99ec037e8548/test/e2e/fixture/app/actions.go#L450-L461) acknowledges it adding the 3 seconds sleep ( and the comment ofc )

**Comment by agaudreault** [comment]
Will close this PR as I don't think it fixes the linked issue which needs clarifications.
